design/batch_retrieval_design.md

```diff
@@ -0,0 +1,20 @@
+# Batch retrieval design
+
+This document will cover the options for enhancing Batch Retrieval requests.
+
+### Motivation
+
+*Previously*, for the init container solution, each Secrets Provider was sitting within the same Deployment as the app. Because of this, when a batch retrieval request was made, the requests were dispersed across many call to the Conjur server. If one K8s Secret was not able to be updated with a Conjur value (for example due to a permission error), then the batch request would fail and only that pod would not spin up successfully. 
+
+*Now* that the Secrets Provider is outside, a single batch retrieval request will be made to Conjur for *all* application. So if there is a failure in receiving one of the values, then the whole request would fail. In other words, **no application** will spin up in the namespace.
+
+|      | Solution                                                     | Pros                                                         | Cons                                                         | Effort (T-shirt size) |
+| ---- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | --------------------- |
+| 1    | *Client side:* <br />For each K8s Secret, perform a new Batch request | - No server side changes / no breaking changes               | Load on server with extra calls (*)                          | S, 5 days             |
+| 2    | *Server side:*<br />Update Batch retrieval to return list of variables **and** their response (success/failure) (**) | - Will help us during rotation for Milestone 2<br />- Better / straightforward design for how batch endpoints | - Requires both client/server implementation<br />- Need to handle backwards compatibility | M, 10 days            |
```

Comment:
I think we should update batch retrieval on the server side. Keep the current behavior unless a specific query parameter is provided. Such as `ignore` or `continueOnFailure`.

The current endpoint looks like:
```
/secrets?variable_ids=myorg:variable:secret1,myorg:variable:secret1
```

The new endpoint should look like:
```
/secrets?continueOnError=true&variable_ids=myorg:variable:secret1,myorg:variable:secret1
```

On the `secrets-provider-for-k8s` side it should be optional. Such as an environment variable `CONJUR_IGNORE_UNSUCCESSFUL_PASSWORD_RETRIEVAL=true` or something like that.

Reply:
We should consider here how the reply of the request looks currently

Reply:
@AndrewCopeland this is could definitely be an option and might save us time.

@abrahamko ^